### ISSUE TITLE
Crashes with STRICT_MODULE_RWX

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#374]
Laurent reports his VM crashes intermittently with STRICT_MODULE_RWX enabled.

https://lore.kernel.org/linuxppc-dev/f7624d58-80e1-6912-1867-7874f1a569f5@redhat.com/

```
[    8.496850] kernel tried to execute exec-protected page (c008000004073278) - exploit attempt? (uid: 0)
[    8.496921] BUG: Unable to handle kernel instruction fetch
[    8.496954] Faulting instruction address: 0xc008000004073278
[    8.496994] Oops: Kernel access of bad area, sig: 11 [#1]
[    8.497028] LE PAGE_SIZE=64K MMU=Radix SMP NR_CPUS=2048 NUMA pSeries
[    8.497071] Modules linked in: drm virtio_console fuse drm_panel_orientation_quirks ...
[    8.497186] CPU: 3 PID: 44 Comm: kworker/3:1 Not tainted 5.14.0-rc4+ #12
[    8.497228] Workqueue: events control_work_handler [virtio_console]
[    8.497272] NIP:  c008000004073278 LR: c008000004073278 CTR: c0000000001e9de0
[    8.497320] REGS: c00000002e4ef7e0 TRAP: 0400   Not tainted  (5.14.0-rc4+)
[    8.497361] MSR:  800000004280b033 <SF,VEC,VSX,EE,FP,ME,IR,DR,RI,LE>  CR: 24002822 XER: 200400cf
[    8.497426] CFAR: c0000000001e9e44 IRQMASK: 1
[    8.497426] GPR00: c008000004073278 c00000002e4efa80 c000000002a26b00 c000000042c39520
[    8.497426] GPR04: 0000000000000001 0000000000000000 0000000000000000 00000000000000ff
[    8.497426] GPR08: 0000000000000001 c000000042c39520 0000000000000001 c008000004076008
[    8.497426] GPR12: c0000000001e9de0 c0000001fffccb00 c00000000018ba88 c00000002c91d400
[    8.497426] GPR16: 0000000000000000 0000000000000000 0000000000000000 0000000000000000
[    8.497426] GPR20: 0000000000000000 0000000000000000 0000000000000000 c008000004080340
[    8.497426] GPR24: c0080000040a01e8 0000000000000000 0000000000000000 c00000002e0975c0
[    8.497426] GPR28: c00000002ce72940 c000000042c39520 0000000000000048 0000000000000038
[    8.497891] NIP [c008000004073278] fill_queue+0xf0/0x210 [virtio_console]
[    8.497934] LR [c008000004073278] fill_queue+0xf0/0x210 [virtio_console]
[    8.497976] Call Trace:
[    8.497993] [c00000002e4efa80] [c00800000407323c] fill_queue+0xb4/0x210 [virtio_console] (unreliable)
[    8.498052] [c00000002e4efae0] [c008000004073a90] add_port+0x1a8/0x470 [virtio_console]
[    8.498102] [c00000002e4efbb0] [c0080000040750f4] control_work_handler+0xbc/0x1e8 [virtio_console]
[    8.498160] [c00000002e4efc60] [c00000000017f4f0] process_one_work+0x290/0x590
[    8.498212] [c00000002e4efd00] [c00000000017f878] worker_thread+0x88/0x620
[    8.498256] [c00000002e4efda0] [c00000000018bc14] kthread+0x194/0x1a0
[    8.498299] [c00000002e4efe10] [c00000000000cf54] ret_from_kernel_thread+0x5c/0x64
[    8.498349] Instruction dump:
[    8.498374] 7da96b78 a14d0c8a 419c00b0 2f8a0000 419eff88 b32d0c8a 7c0004ac 4bffff7c
[    8.498430] 60000000 60000000 7fa3eb78 48002d95 <e8410018> 38600000 480025e1 e8410018
[    8.498485] ---[ end trace 16ee10903290b647 ]---

mr      r9,r13
lhz     r10,3210(r13)
blt     cr7,0xb0
cmpwi   cr7,r10,0
beq     cr7,0xffffffffffffff88
sth     r25,3210(r13)
hwsync
b       0xffffffffffffff7c
nop
nop
mr      r3,r29
bl      0x2d94
ld      r2,24(r1)	<---
```

Bizarrely we are faulting when returning to a page we have previously executed from.
We shouldn't be changing permissions of the module while executing it.

**Comment by iamjpn** [comment]
Updates from Fabiano  https://lore.kernel.org/linuxppc-dev/8735rdt0i4.fsf@linux.ibm.com/
and myself https://lore.kernel.org/linuxppc-dev/CACzsE9pLgn+w=Ohx3gSaMxnZk8E2itmBPy_V2NVPrXjqL=Ef+Q@mail.gmail.com/

The comment above `page_change_attr()` reads

```
 * Updates the attributes of a page in three steps:
 *
 * 1. invalidate the page table entry
 * 2. flush the TLB
 * 3. install the new entry with the updated attributes
 *
 * Invalidating the pte means there are situations where this will not work
 * when in theory it should.
 * For example:
 * - removing write from page whilst it is being executed
 * - setting a page read-only whilst it is being read by another CPU
 ```

In the module loading code there is a situation where we remove write from a page whilst it is being executed so this doesn't work.

**Comment by mpe** [comment]
v1 [patch and discussion](https://lore.kernel.org/linuxppc-dev/20210817132552.3375738-1-mpe@ellerman.id.au/).
v2 [patch](https://lore.kernel.org/linuxppc-dev/20210818120518.3603172-1-mpe@ellerman.id.au/)

Fixed in torvalds/linux@9f7853d7609d.
